pkg/providers/instancetype/offering/provider.go

```diff
@@ -0,0 +1,274 @@
+	for _, it := range instanceTypes {
+		offerings := p.createOfferings(
+			it,
+			nodeClass,
+			allZones,
+			subnetZones,
+			snapshotUUID,
+		)
+		for _, of := range offerings {
+			InstanceTypeOfferingAvailable.Set(float64(lo.Ternary(of.Available, 1, 0)), map[string]string{
+				instanceTypeLabel: it.Name,
+				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
+				zoneLabel:         of.Requirements.Get(corev1.LabelTopologyZone).Any(),
+			})
+			InstanceTypeOfferingPriceEstimate.Set(of.Price, map[string]string{
+				instanceTypeLabel: it.Name,
+				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
+				zoneLabel:         of.Requirements.Get(corev1.LabelTopologyZone).Any(),
+			})
+		}
+
+		its = append(its, &cloudprovider.InstanceType{
+			Name:         it.Name,
+			Requirements: it.Requirements,
+			Offerings:    offerings,
+			Capacity:     it.Capacity,
+			Overhead:     it.Overhead,
+		})
+	}
+	return its
+}
+
+//nolint:gocyclo
+func (p *DefaultProvider) createOfferings(
+	it *cloudprovider.InstanceType,
+	nodeClass *v1.EC2NodeClass,
+	allZones sets.Set[string],
+	subnetZones map[string]string,
+	snapshotUUID *types.UID,
+) cloudprovider.Offerings {
+	itZones := sets.New(it.Requirements.Get(corev1.LabelTopologyZone).Values()...)
+
+	offerings := []cloudprovider.Offering{}
+	for zone := range allZones {
+		for _, capacityType := range it.Requirements.Get(karpv1.CapacityTypeLabelKey).Values() {
+			// Reserved capacity types are constructed separately, skip them for now.
+			if capacityType == karpv1.CapacityTypeReserved {
+				continue
+			}
+
+			isUnavailable := p.unavailableOfferings.IsUnavailable(it.Name, zone, capacityType)
+			_, hasSubnetZone := subnetZones[zone]
+			var price float64
+			var hasPrice bool
+			switch capacityType {
+			case karpv1.CapacityTypeOnDemand:
+				price, hasPrice = p.pricingProvider.OnDemandPrice(ec2types.InstanceType(it.Name))
+			case karpv1.CapacityTypeSpot:
+				price, hasPrice = p.pricingProvider.SpotPrice(ec2types.InstanceType(it.Name), zone)
+			default:
+				panic(fmt.Sprintf("invalid capacity type %q in requirements for instance type %q", capacityType, it.Name))
+			}
+			offering := cloudprovider.Offering{
+				Requirements: scheduling.NewRequirements(
+					scheduling.NewRequirement(karpv1.CapacityTypeLabelKey, corev1.NodeSelectorOpIn, capacityType),
+					scheduling.NewRequirement(corev1.LabelTopologyZone, corev1.NodeSelectorOpIn, zone),
+					scheduling.NewRequirement(v1.LabelCapacityReservationID, corev1.NodeSelectorOpDoesNotExist),
+				),
+				Price:     price,
+				Available: !isUnavailable && hasPrice && itZones.Has(zone) && hasSubnetZone,
+			}
+			if id, ok := subnetZones[zone]; ok {
+				offering.Requirements.Add(scheduling.NewRequirement(v1.LabelTopologyZoneID, corev1.NodeSelectorOpIn, id))
+			}
+			offerings = append(offerings, offering)
+		}
+	}
+
+	for i := range nodeClass.Status.CapacityReservations {
```

Comment:
Just validating: We _have_ to create separate offerings for each reservation because we could have cross-over on reservations between different NodePools -- if we decrement one from one reservation on one NodePool, we need to know that that will decrement for the other NodePool as well -- do I have that right?

Reply:
Yeah, if it weren't for the cross-nodepool consideration we would be able to pool them together.